src/mono/mono/eglib/giconv.c

```diff
@@ -559,162 +321,92 @@ g_utf8_to_ucs4_fast (const gchar *str, glong len, glong *items_written)
 	return outbuf;
 }
 
-static gunichar2 *
-eg_utf8_to_utf16_general (const gchar *str, glong len, glong *items_read, glong *items_written, gboolean include_nuls, gboolean replace_invalid_codepoints, GCustomAllocator custom_alloc_func, gpointer custom_alloc_data, GError **err, unsigned endian)
+static FORCE_INLINE (void)
+map_error(GError **err)
 {
-	gunichar2 *outbuf, *outptr;
-	size_t outlen = 0;
-	size_t inleft;
-	char *inptr;
-	gunichar c;
-	int u, n;
-
-	g_return_val_if_fail (str != NULL, NULL);
-
-	if (len < 0) {
-		if (include_nuls) {
-			g_set_error (err, G_CONVERT_ERROR, G_CONVERT_ERROR_FAILED, "Conversions with embedded nulls must pass the string length");
-			return NULL;
-		}
-
-		len = (glong)strlen (str);
-	}
-
-	inptr = (char *) str;
-	inleft = len;
-
-	while (inleft > 0) {
-		if ((n = decode_utf8 (inptr, inleft, &c)) < 0)
-			goto error;
-
-		if (c == 0 && !include_nuls)
-			break;
-
-		if ((u = g_unichar_to_utf16_endian (c, NULL, endian)) < 0) {
-			if (replace_invalid_codepoints) {
-				u = 2;
-			} else {
-				mono_set_errno (EILSEQ);
-				goto error;
-			}
-		}
-
-		outlen += u;
-		inleft -= n;
-		inptr += n;
-	}
-
-	if (items_read)
-		*items_read = GPTRDIFF_TO_LONG (inptr - str);
-
-	if (items_written)
-		*items_written = (glong)outlen;
-
-	if (G_LIKELY (!custom_alloc_func))
-		outptr = outbuf = g_malloc ((outlen + 1) * sizeof (gunichar2));
-	else
-		outptr = outbuf = (gunichar2 *)custom_alloc_func ((outlen + 1) * sizeof (gunichar2), custom_alloc_data);
-
-	if (G_UNLIKELY (custom_alloc_func && !outbuf)) {
-		mono_set_errno (ENOMEM);
-		goto error;
-	}
-
-	inptr = (char *) str;
-	inleft = len;
-
-	while (inleft > 0) {
-		if ((n = decode_utf8 (inptr, inleft, &c)) < 0)
-			break;
-
-		if (c == 0 && !include_nuls)
-			break;
-
-		u = g_unichar_to_utf16_endian (c, outptr, endian);
-		if ((u < 0) && replace_invalid_codepoints) {
-			outptr[0] = 0xFFFD;
-			outptr[1] = 0xFFFD;
-			u = 2;
-		}
-
-		outptr += u;
-		inleft -= n;
-		inptr += n;
-	}
-
-	*outptr = '\0';
-
-	return outbuf;
-
-error:
-	if (errno == ENOMEM) {
+	if (errno == 0) return;
+	if (errno == MINIPAL_ERROR_INSUFFICIENT_BUFFER) {
 		g_set_error (err, G_CONVERT_ERROR, G_CONVERT_ERROR_NO_MEMORY,
 			     "Allocation failed.");
-	} else if (errno == EILSEQ) {
-		g_set_error (err, G_CONVERT_ERROR, G_CONVERT_ERROR_ILLEGAL_SEQUENCE,
-			     "Illegal byte sequence encountered in the input.");
-	} else if (items_read) {
-		/* partial input is ok if we can let our caller know... */
 	} else {
 		g_set_error (err, G_CONVERT_ERROR, G_CONVERT_ERROR_PARTIAL_INPUT,
```

Comment:
What can we hit this case? I see that we only ever set `errno` to `MINIPAL_ERROR_INSUFFICIENT_BUFFER`.

Reply:
Since conversion to assets, this branch is no longer needed.  Deleted.